### PR TITLE
Prevent cannibalism on transition to multicell

### LINKED ISSUE #3155: Prevent unintentional cannibalism when species type changes to multicellular
cannibalism is possible while there exists cells of the player's previous species as the microbe species instance is not equal to the early multicellular species instance

### PR DESCRIPTION
**Pull request by astral-grayson**
**Brief Description of What This PR Does**

On changing to early multicellular mode, generate an array of all existing microbes of the player's species, and apply the new multicellular species to them. This prevents cannibalism and the immediate death of the player to their former colony.

**Related Issues**

Closes #3155

**Progress Checklist**

Note: before starting this checklist the PR should be marked as non-draft.

- [x] PR author has checked that this PR works as intended and doesn't
      break existing features:
      https://wiki.revolutionarygamesstudio.com/wiki/Testing_Checklist
      (this is important as to not waste the time of Thrive team
      members reviewing this PR)
- [x] Initial code review passed (this and further items should not be checked by the PR author)
- [ ] Functionality is confirmed working by another person (see above checklist link)
- [ ] Final code review is passed and code conforms to the 
      [styleguide](https://github.com/Revolutionary-Games/Thrive/blob/master/doc/style_guide.md).

Before merging all CI jobs should finish on this PR without errors, if
there are automatically detected style issues they should be fixed by
the PR author. Merging must follow our
[styleguide](https://github.com/Revolutionary-Games/Thrive/blob/master/doc/style_guide.md#git).

### CODE REVIEW

**Comment by hhyyrylainen** [comment]
I'm unsure about this approach... this does seem pretty elegant way to get around the species objects not comparing equally. I was thinking that I would fix this issue myself by overloading .Equals for the Species types to just compare IDs, only caveat with that is that quite many comparisons need to be found and updated, so this approach gets around that.
I was going to comment my worry that converting the existing cells to multicellular will allow them to start growing colonies, but maybe that isn't the case as the species only has a single cell when applied? But then again when exiting the editor the species referenced by the cells will have additional cells in the layout.

Anyone else have thoughts on this? I'm sort of leaning towards accepting this approach as the one needing less overall code changes.

**Comment by astral-grayson** [comment]
When you exit the editor, each microbe is in the basic single cell bud state. I feel like having the single cell and multicell versions evaluate equal would make less sense over time as the player species diverges from the single celled form with successive evolutions. Maybe we should be explicit about the difference and turn the old player species into a branch with a different name? I also entertained the idea of just despawning some or all of the nearby player microbes, if we don't want that many copies of the player's species around.

**Comment by hhyyrylainen** [comment]
>  sense over time as the player species diverges from the single celled form with successive evolutions.

The single celled version of the player species is removed immediately when the player becomes multicellular. The earlier species hanging around is a bug.

Related bug that I freshly opened when noticing that the old species still kept spawning (which is not intended): https://github.com/Revolutionary-Games/Thrive/issues/3274

File: src/microbe_stage/Microbe.cs [hunk 1]
```diff
@@ -395,6 +395,9 @@ public void ApplySpecies(Species species)
             MulticellularCellType = earlyMulticellularSpecies.Cells[0].CellType;
 
             cachedMulticellularSpecies = earlyMulticellularSpecies;
+
+            if (IsPlayerMicrobe)
+                GD.Print("Applying multicellular species type to player");
```
Comment (hhyyrylainen) [inline src/microbe_stage/Microbe.cs:400]: This print is not necessary. If you want to confirm that the player also gets applied the new species add a check in MoveToMulticellular that prints an error if the player wasn't processed.

File: src/microbe_stage/MicrobeStage.cs [hunk 2]
```diff
@@ -716,6 +720,24 @@ public void OnFinishTransitioning()
             TutorialEventType.EnteredMicrobeStage, new CallbackEventArgs(HUD.PopupPatchInfo), this);
     }
 
+    /// <summary>
+    ///   Helper function for transition to multicellular
+    ///   Returns array of all microbes of Player's species
```
Comment (hhyyrylainen) [inline src/microbe_stage/MicrobeStage.cs:725]: This line could be a separate `<returns>` xml block

File: src/microbe_stage/MicrobeStage.cs [hunk 3]
```diff
@@ -716,6 +720,24 @@ public void OnFinishTransitioning()
...
+    private Microbe[] GetAllPlayerSpeciesMicrobes()
+    {
+        if (Player == null)
+        {
+            GD.PrintErr("Could not get player species microbes: no Player node");
+            return new Microbe[0];
+        }
+
+        var microbes = rootOfDynamicallySpawned.GetTree().GetNodesInGroup(Constants.AI_TAG_MICROBE).Cast<Microbe>()
+            .ToArray();
```
Comment (hhyyrylainen) [inline src/microbe_stage/MicrobeStage.cs:736]: This ToArray is also completely unnecessary, you can just chain LINQ as much as you want (well within reason, we do have a styleguide which doesn't recommend infinite LINQ).

**Review by hhyyrylainen — APPROVED** [review]
Did the refactoring I wanted so that I can merge this in time for 0.5.8.1

**Comment by astral-grayson** [comment]
Cool, thanks :) I spent the day on my coursework instead, haha.

**Pull request merged**